src/backends/oracle/standard-into-type.cpp

```diff
@@ -235,41 +235,58 @@ void oracle_standard_into_type_backend::pre_fetch()
 void oracle::read_from_lob(oracle_session_backend& session,
     OCILobLocator * lobp, std::string & value)
 {
-    ub4 len;
-
-    sword res = OCILobGetLength(session.svchp_, session.errhp_,
-        lobp, &len);
+    // We can't get the CLOB size in bytes directly, only in characters, which
+    // is useless for UTF-8 as it doesn't tell us how much memory do we
+    // actually need for storing it, so we'd have to allocate 4 bytes for every
+    // character which could be a huge overkill. So instead read the CLOB in
+    // chunks of its natural size until we get everything.
+    ub4 len = 0;
+    sword res = OCILobGetChunkSize(session.svchp_, session.errhp_, lobp, &len);
     if (res != OCI_SUCCESS)
     {
         throw_oracle_soci_error(res, session.errhp_);
     }
 
-    std::vector<char> buf(len);
+    value.clear();
+
+    if (!len)
+        return;
 
-    if (len != 0)
+    // Read the LOB in chunks into the buffer while anything remains to be read.
+    for (bool done = false; !done; )
     {
-        ub4 lenChunk = len;
-        ub4 offset = 1;
-        do
+        auto const prevSize = value.size();
+        value.resize(prevSize + len);
```

Comment:
Could it be that "reserve" should be used here?

Reply:
No, accessing the buffer wouldn't be valid without `resize()`.

Reply:
Then we have resize being called twice in a row on the second and subsequent iterations. See line 284

Reply:
Yes, but this is correct and intentional: we first resize to get a big enough buffer and then resize to the actual length. The second resize won't allocate any memory (and may, in fact, do nothing at all, if `lenChunk == len`).

What do you think is a problem?

Reply:
Yes, but I see nothing wrong here. If you do, please explain what.

I have no idea why does Oracle return us only 2000 bytes when we told it to get as much as possible that fits in 4000, but I don't suppose we can change how it behaves.

Reply:
May be try so:
``

    // Read the LOB in chunks into the buffer while anything remains to be read.
    auto theUsedSize = value.size();
    do 
    {
        // resize to max possible read size
        value.resize( theUsedSize + len );

        // By setting the input length to 0, we tell Oracle to read as many
        // bytes as possible (so called "streaming" mode).
        ub4 lenChunk = 0;
        res = OCILobRead(session.svchp_, session.errhp_, lobp,
                         &lenChunk,
                         1, // Only used for the first chunk, ignored later.
                         const_cast<char*>(value.data()) + theUsedSize, len,
                         0, 0, 0, 0);
        theUsedSize += lenChunk;

        if ( res != OCI_SUCCESS && res != OCI_NEED_DATA )
        {
            throw_oracle_soci_error ( res, session.errhp_ );
        }

    } while ( res == OCI_NEED_DATA );

    // resize to the actual size
    value.resize( theUsedSize );
    // We may have over-allocated the string, especially when a big chunk size
    // is used, so release the unused memory.
    value.shrink_to_fit();

``

Reply:
This would work too AFAICS but seems a bit more difficult to understand to me.

It also results in exactly the same allocation pattern because resizing the string to shrink doesn't allocate (or free) any memory and is a very cheap operation (it basically adjusts one pointer or size, i.e. about the same as `theUsedSize += lenChunk`).

IOW I still have no idea what problem does this solve?

Reply:
No problem at all, but my approach seems a bit more readable to me and potentially more efficient. However, performance is not crucial in this case. I’m perfectly fine with your solution as well